managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java

```diff
@@ -472,8 +484,87 @@ void close(ManagedLedger ledger) {
         entryCacheManager.removeEntryCache(ledger.getName());
     }
 
+    @Override
+    public void shutdownGracefully() throws ManagedLedgerException, InterruptedException {
+        if (closed) {
+            throw new ManagedLedgerException.ManagedLedgerFactoryClosedException();
+        }
+        closed = true;
+
+        statsTask.cancel(true);
+        flushCursorsTask.cancel(true);
+
+        int closedLedgerNum = 0;
+
+        while (!ledgers.isEmpty()) { //use while loop for ledgers added during shutdown.
+            List<String> ledgerNames = new ArrayList<>(this.ledgers.keySet());
+            int numLedgers = ledgerNames.size();
+            final CountDownLatch latch = new CountDownLatch(numLedgers);
```

Comment:
How about make the `shutdownGracefully()` asynchronously?

Reply:
Great idea, shutdownAsync is more flexible.